yapapi/events.py

```diff
@@ -107,13 +107,37 @@
 
 
 #   ABSTRACT EVENTS
-@attr.s
+@attr.s(frozen=True, repr=False)
 class Event(abc.ABC):
     """An abstract base class for types of events emitted by `Executor.submit()`."""
 
     exc_info: Optional[ExcInfo] = attr.ib(default=None, kw_only=True)
     """Tuple containing exception info as returned by `sys.exc_info()`, if applicable."""
 
+    def __str__(self) -> str:
+        """Mimics Python's default `repr` format, but excludes the field `exc_info` from it.
+
+        If `exc_info` is not `None`, its underlying exception is included in the result string
+        under the key `exception`.
+        """
+        fields: Tuple[attr.Attribute] = attr.fields(self.__class__)  # type: ignore
+        field_reprs = ""
```

Comment:
I prefer to gather parts in an array.
```
field_reprs = []
for ...
    field_reprs.append(repr(...))
return field_reprs.join(', ')
```
-> this way we don't have ugly `rstrip` :)

(I'm not saying this is important in any way)